nncf/scopes.py

```diff
@@ -130,78 +130,142 @@ def convert_ignored_scope_to_list(ignored_scope: Optional[IgnoredScope]) -> List
+) -> Tuple[IgnoredScope, Dict[str, Set[str]]]:
+    """
+    Returns matched ignored scope for provided graphs along with all found matches.
+    The resulted ignored scope consist of all matched rules.
+    The found matches consist of a ditionary with a rule name as a key and matched node names as a value.
+
+    :param ignored_scope: Ignored scope instance.
+    :param nncf_graphs: Graphs.
+    :returns: Mathced ignored scope along with all matches.
+    """
+    names, patterns, types, subgraphs_numbers = set(), set(), set(), set()
+    matches = {"names": set(), "patterns": set(), "types": set(), "subgraphs": set()}
+
+    for graph in nncf_graphs:
+        node_names = set(node.node_name for node in graph.nodes.values())
+
+        for ignored_node_name in filter(lambda name: name in node_names, ignored_scope.names):
+            matches["names"].add(ignored_node_name)
+            names.add(ignored_node_name)
+
+        for str_pattern in ignored_scope.patterns:
+            pattern = re.compile(str_pattern)
+            pattern_matched_names = set(filter(pattern.match, node_names))
+            if pattern_matched_names:
+                matches["patterns"].update(pattern_matched_names)
+                patterns.add(str_pattern)
+
+        for node in graph.get_nodes_by_types(set(ignored_scope.types)):
+            matches["types"].add(node.node_name)
+            types.add(node.node_type)
+
+        for i, subgraph in enumerate(ignored_scope.subgraphs):
+            names_from_subgraph = get_ignored_node_names_from_subgraph(graph, subgraph)
+            if names_from_subgraph:
+                matches["subgraphs"].update(names_from_subgraph)
+                subgraphs_numbers.add(i)
+
+    matched_ignored_scope = IgnoredScope(
+        names=list(names),
+        patterns=list(patterns),
+        types=list(types),
+        subgraphs=[subgraph for i, subgraph in enumerate(ignored_scope.subgraphs) if i in subgraphs_numbers],
+        validate=ignored_scope.validate,
+    )
+    return matched_ignored_scope, matches
+
+
+def get_unmatched_ignored_scope(matched_ignored_scope: IgnoredScope, ignored_scope: IgnoredScope) -> IgnoredScope:
+    """
+    Returns unmatched ignored scope rules from full ignored scope and matched ignored scope.
+
+    :param matched_ignored_scope: Matched ingored scope.
+    :param ignored_scope: Full ignored scope.
+    :return: Unmatched ignored scope.
+    """
+    assert matched_ignored_scope.validate == ignored_scope.validate
+    return IgnoredScope(
+        names=[name for name in ignored_scope.names if name not in matched_ignored_scope.names],
+        patterns=[pattern for pattern in ignored_scope.patterns if pattern not in matched_ignored_scope.patterns],
+        types=[type for type in ignored_scope.types if type not in matched_ignored_scope.types],
+        subgraphs=[subgraph for subgraph in ignored_scope.subgraphs if subgraph not in matched_ignored_scope.subgraphs],
```

Comment:
Seems that it is possible to avoid these `for`'s if we would collect `unmatched` items during `get_ignored_scope_match`.

Reply:
From a design perspective, I don't think that it is a good idea to add an extra, sometimes not used, return variable, which can be calculated from the already returned variable using this method `get_unmatched_ignored_scope(...)`.
The dummy example: To create a function that returns all even numbers from 0..N. Do you think there is a reason to add a return value of odd numbers?

Reply:
The `get_unmatched_ignored_scope` looks like an important part of the new approach. So, I don't understand why can't we avoid extra calculations for this method and reorganize the code better.

Reply:
See the previous comment.